lib/maestro/aggregate/root.ex

```diff
@@ -0,0 +1,367 @@
+  context.
+  """
+  @callback call(id(), msg :: any()) :: any()
+
+  @optional_callbacks initial_state: 0, prepare_snapshot: 1, use_snapshot: 2
+
+  def child_spec(opts) do
+    %{
+      id: __MODULE__,
+      start: {__MODULE__, :start_link, [opts]},
+      type: :worker,
+      restart: :temporary,
+      shutdown: 500
+    }
+  end
+
+  def start_link(opts) do
+    mod = Keyword.fetch!(opts, :module)
+    id = Keyword.fetch!(opts, :aggregate_id)
+    mod.start_link(id)
+  end
+
+  @doc false
+  def create_aggregate(agg_id, mod, command_prefix, event_prefix) do
+    %__MODULE__{
+      id: agg_id,
+      sequence: 0,
+      state: mod.initial_state(),
+      module: mod,
+      command_prefix: command_prefix,
+      event_prefix: event_prefix
+    }
+  end
+
+  @doc false
+  def update_aggregate(agg, max_seq \\ Store.max_sequence()) do
+    # from latest snapshot
+    agg =
+      case Store.get_snapshot(
+             agg.id,
+             agg.sequence,
+             max_sequence: max_seq
+           ) do
+        nil ->
+          agg
+
+        %Snapshot{} = snap ->
+          %{
+            agg
+            | state: agg.module.use_snapshot(agg, snap),
+              sequence: snap.sequence
+          }
+      end
+
+    # plus trailing events
+    events = Store.get_events(agg.id, agg.sequence, max_sequence: max_seq)
+    apply_events(agg, events)
+  end
+
+  @doc false
+  def hydrate(agg, target_seq) do
+    agg.id
+    |> create_aggregate(agg.module, agg.command_prefix, agg.event_prefix)
+    |> update_aggregate(target_seq)
+    |> Map.get(:state)
+  end
+
+  @doc false
+  def eval_command(agg, command) do
+    with seq <- Map.get(command, :sequence, 0),
+         agg <- update_aggregate(agg, seq),
+         module <- lookup_module(agg.command_prefix, command.type),
+         events <- module.eval(agg, command) do
+      persist_events(agg, command, events)
+    end
+  end
+
+  @doc false
+  def persist_snapshot(snapshot) do
+    Store.commit_snapshot(snapshot)
+  end
+
+  defp persist_events(agg, command, events) do
+    case Store.commit_events(events) do
+      :ok ->
+        apply_events(agg, events)
+
+      {:error, :retry_command} ->
+        agg
+        |> update_aggregate()
+        |> eval_command(command)
+    end
+  end
+
+  @doc false
+  def apply_events(agg, []), do: agg
+
+  @doc false
+  def apply_events(agg, events) do
+    state =
+      Enum.reduce(events, agg.state, fn event, state ->
+        module = lookup_module(agg.event_prefix, event.type)
+        module.apply(state, event)
+      end)
+
+    %{agg | state: state, sequence: max_seq(events)}
+  end
+
+  @doc false
+  def to_snapshot(agg, body) do
+    %Snapshot{
+      aggregate_id: agg.id,
+      sequence: agg.sequence,
+      body: body
+    }
+  end
+
+  @doc false
+  def lookup_module(prefix, type) do
+    name =
+      type
+      |> String.split(".")
+      |> Enum.map_join(".", &Macro.camelize/1)
```

Comment:
Since the library now owns the event/command type conventions, it might be good to move the inverse of this function into meastro (the event_type function in core). Since you have the prefix here, this would remove an assumption about the number of elements in the module name that is hard-coded in core.

Reply:
That makes sense. I had no direct need of it in the sample aggregate, but it's also extremely boiled down. Thanks!